File: mcp_client.py
import asyncio
import json
import subprocess
from typing import Any, Dict, List, Optional
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from rich.console import Console

console = Console()


class PlaywrightMCPClient:
    def __init__(self, command: str, args: List[str]):
        self.command = command
        self.args = args
        self.session: Optional[ClientSession] = None
        self.read = None
        self.write = None
        self.tools: List[Dict[str, Any]] = []
        
    async def __aenter__(self):
        await self.connect()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.disconnect()
        
    async def connect(self):
        console.print("[cyan]Starting Playwright MCP server...[/cyan]")
        
        server_params = StdioServerParameters(
            command=self.command,
            args=self.args,
            env=None
        )
        
        stdio_context = stdio_client(server_params)
        self.read, self.write = await stdio_context.__aenter__()
        self.session = ClientSession(self.read, self.write)
        
        await self.session.__aenter__()
        await self.session.initialize()
        
        tools_list = await self.session.list_tools()
        self.tools = [tool.model_dump() for tool in tools_list.tools]
        
        console.print(f"[green]✓ MCP server connected. Available tools: {len(self.tools)}[/green]")
        
        return self
    
    async def disconnect(self):
        if self.session:
            await self.session.__aexit__(None, None, None)
            console.print("[cyan]MCP server disconnected[/cyan]")
    
    async def call_tool(self, tool_name: str, arguments: Dict[str, Any]) -> Any:
        if not self.session:
            raise RuntimeError("MCP client not connected")
        
        console.print(f"[yellow]→ Calling tool: {tool_name}[/yellow]")
        console.print(f"[dim]  Arguments: {json.dumps(arguments, indent=2)}[/dim]")
        
        result = await self.session.call_tool(tool_name, arguments)
        
        console.print(f"[green]✓ Tool executed: {tool_name}[/green]")
        
        return result
    
    def get_tools_description(self) -> str:
        if not self.tools:
            return "No tools available"
        
        descriptions = []
        for tool in self.tools:
            tool_desc = f"- {tool['name']}: {tool.get('description', 'No description')}"
            if 'inputSchema' in tool:
                props = tool['inputSchema'].get('properties', {})
                if props:
                    params = ", ".join(props.keys())
                    tool_desc += f"\n  Parameters: {params}"
            descriptions.append(tool_desc)
        
        return "\n".join(descriptions)

File: src/mcp-client.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import chalk from 'chalk';

export interface MCPTool {
  name: string;
  description?: string;
  inputSchema?: {
    type: string;
    properties?: Record<string, unknown>;
    required?: string[];
  };
}

export class PlaywrightMCPClient {
  private client: Client | null = null;
  private transport: StdioClientTransport | null = null;
  private tools: MCPTool[] = [];

  async connect(headless: boolean = false): Promise<void> {
    console.log(chalk.gray('Connecting to Playwright MCP server...'));

    // Use Playwright's bundled chromium, no external path
    const args = [
      '@playwright/mcp@latest',
      '--browser',
      'chromium'
    ];

    if (headless) {
      args.push('--headless');
    }

    this.transport = new StdioClientTransport({
      command: 'npx',
      args,
      env: {
        ...process.env,
        XDG_CONFIG_HOME: process.env.XDG_CONFIG_HOME || '/tmp/.config',
        HOME: process.env.HOME || '/tmp',

        // Allow Playwright MCP to download browsers normally
        PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD: '0',
      },
    });

    this.client = new Client(
      { name: 'playwright-mcp-cli', version: '1.0.0' },
      { capabilities: {} }
    );

    await this.client.connect(this.transport);
    console.log(chalk.green('✓ Connected to Playwright MCP server'));

    const toolsResult = await this.client.listTools();
    this.tools = toolsResult.tools as MCPTool[];
    console.log(chalk.gray(`Available tools: ${this.tools.map(t => t.name).join(', ')}`));
  }

  async disconnect(): Promise<void> {
    if (this.transport) {
      await this.transport.close();
      this.client = null;
      this.transport = null;
      console.log(chalk.gray('Disconnected from Playwright MCP server'));
    }
  }

  getTools(): MCPTool[] {
    return this.tools;
  }

  getToolsDescription(): string {
    return this.tools.map(tool => {
      const params = tool.inputSchema?.properties
        ? Object.entries(tool.inputSchema.properties)
            .map(([key, value]) => `${key}: ${(value as { type?: string }).type || 'any'}`)
            .join(', ')
        : 'no parameters';
      return `- ${tool.name}: ${tool.description || 'No description'} (${params})`;
    }).join('\n');
  }

  async callTool(name: string, args: Record<string, unknown>): Promise<unknown> {
    if (!this.client) {
      throw new Error('Not connected to MCP server');
    }

    console.log(chalk.cyan(`→ Calling tool: ${name}`));
    console.log(chalk.gray(`  Args: ${JSON.stringify(args)}`));

    try {
      const result = await this.client.callTool({ name, arguments: args });

      if (result.content && Array.isArray(result.content)) {
        for (const item of result.content) {
          if (item.type === 'text') {
            console.log(
              chalk.green(
                `  ✓ Result: ${(item as { text: string }).text.slice(0, 200)}${
                  (item as { text: string }).text.length > 200 ? '...' : ''
                }`
              )
            );
          } else if (item.type === 'image') {
            console.log(chalk.green(`  ✓ Image captured`));
          }
        }
      }

      return result;
    } catch (error) {
      console.error(
        chalk.red(
          `  ✗ Tool call failed: ${
            error instanceof Error ? error.message : error
          }`
        )
      );
      throw error;
    }
  }
}
